### PR TITLE
HIVE-21742 : Vectorization: CASE result type casting

### PR DESCRIPTION
**Pull request by vineetgarg02**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/RexNodeConverter.java [hunk 1]
```diff
@@ -571,7 +571,7 @@ private RexNode handleExplicitCast(ExprNodeGenericFuncDesc func, List<RexNode> c
    * If a CASE has branches with string/int/boolean branch types; there is no common type.
    */
   private List<RexNode> adjustCaseBranchTypes(List<RexNode> nodes, RelDataType retType) {
-    List<RelDataType> branchTypes = new ArrayList<>();
```
Comment (t3rmin4t0r) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/RexNodeConverter.java]: Commenting out code or removing it?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java [hunk 2]
```diff
@@ -1274,6 +1274,40 @@ protected ExprNodeDesc getXpathOrFuncExprNodeDesc(ASTNode expr,
               children);
         }
 
+        // introduce cast for expressions which are of different type than parent WHEN expr's type
+        // this is done so that vectorization is able to vectorize appropriately
+        if (genericUDF instanceof GenericUDFWhen && desc instanceof ExprNodeGenericFuncDesc) {
+          String castTypeName = desc.getTypeInfo().getTypeName();
+          if (desc.getTypeInfo() instanceof DecimalTypeInfo) {
```
Comment (t3rmin4t0r) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java]: Why is this a special case?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java [hunk 3]
```diff
@@ -1274,6 +1274,40 @@ protected ExprNodeDesc getXpathOrFuncExprNodeDesc(ASTNode expr,
...
+
+          String whenTypeName = desc.getTypeInfo().getTypeName();
+          List<ExprNodeDesc> updatedChildren = new ArrayList<>();
+          if (castFun != null) {
+            GenericUDF castUDF = castFun.getGenericUDF();
+            if (desc.getTypeInfo() instanceof DecimalTypeInfo || desc
+                .getTypeInfo() instanceof VarcharTypeInfo || desc
+                .getTypeInfo() instanceof CharTypeInfo || desc
+                .getTypeInfo() instanceof TimestampLocalTZTypeInfo) {
+              ((SettableUDF) castUDF).setTypeInfo(desc.getTypeInfo());
```
Comment (t3rmin4t0r) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java]: Special case for varchar(1) -> varchar(2) ?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java [hunk 4]
```diff
@@ -1274,6 +1274,40 @@ protected ExprNodeDesc getXpathOrFuncExprNodeDesc(ASTNode expr,
...
+            }
+            for (int i = 0; i < desc.getChildren().size(); i++) {
+              ExprNodeDesc child = children.get(i);
+              ExprNodeDesc newChild = child;
+              if ((i % 2 == 1 || i == desc.getChildren().size() - 1) && !whenTypeName
```
Comment (t3rmin4t0r) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java]: Does typeName have the scale/precision?

File: ql/src/test/results/clientpositive/infer_join_preds.q.out [hunk 5]
```diff
@@ -1187,7 +1187,7 @@ STAGE PLANS:
...
...gecodeid is null) THEN (1L) ELSE (prchargecodeid) END, ',', CASE WHEN (prtypecodeid is null) THEN ('') ELSE (CAST( prtypecodeid AS STRING)) END, ',', CASE WHEN (practsum is null) THEN (CAST( 1 AS decimal(38,20))) ELSE (practsum) END, ',', CASE WHEN (prsequence is null) THEN (1L) ELSE (prsequence) END, ',', CASE WHEN (length(prmodby) is null) THEN ('') ELSE (CAST( prmodby AS STRING)) END, ',', CASE WHEN (prmodtime is null) THEN (TIMESTAMP'2017-12-08 00:00:00') ELSE (prmodtime) END, ',', CASE WHEN (prrmexported is null) THEN (1L) ELSE (prrmexported) END, ',', CASE WHEN (prrmckdel is null) THEN (1L) ELSE (prrmckdel) END, ',', CASE WHEN (slice_status is null) THEN (1) ELSE (slice_status) END, ',', CASE WHEN (role_id is null) THEN (1L) ELSE (role_id) END, ',', CASE WHEN (length(user_lov1) is null) THEN ('') ELSE (CAST( user_lov1 AS STRING)) END, ',', CASE WHEN (length(user_lov2) is null) THEN ('') ELSE (CAST( user_lov2 AS STRING)) END, ',', CASE WHEN (incident_id is null) THEN (1L) ELSE (incident_id) END, ',', CASE WHEN (incident_investment_id is null) THEN (1L) ELSE (incident_investment_id) END, ',', CASE WHEN (odf_ss_actuals is null) THEN (1L) ELSE (odf_ss_actuals) END) (type: string)
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/infer_join_preds.q.out]: is the casts kicking in from 1 -> 1L?

File: ql/src/test/results/clientpositive/llap/multi_count_distinct_null.q.out [hunk 6]
```diff
@@ -73,7 +73,7 @@ STAGE PLANS:
                 outputColumnNames: _col0, _col1, _col2, _col3
                 Statistics: Num rows: 18 Data size: 1628 Basic stats: COMPLETE Column stats: COMPLETE
                 Select Operator
-                  expressions: CASE WHEN (((_col3 = 3L) and _col0 is not null)) THEN (1) ELSE (null) END (type: int), CASE WHEN (((_col3 = 5L) and _col1 is not null)) THEN (1) ELSE (null) END (type: int), CASE WHEN (((_col3 = 6L) and _col2 is not null)) THEN (1) ELSE (null) END (type: int)
+                  expressions: CASE WHEN (((_col3 = 3L) and _col0 is not null)) THEN (1) ELSE (UDFToInteger(null)) END (type: int), CASE WHEN (((_col3 = 5L) and _col1 is not null)) THEN (1) ELSE (UDFToInteger(null)) END (type: int), CASE WHEN (((_col3 = 6L) and _col2 is not null)) THEN (1) ELSE (UDFToInteger(null)) END (type: int)
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/llap/multi_count_distinct_null.q.out]: Cast not being constant folded?

File: ql/src/test/results/clientpositive/llap/subquery_select.q.out [hunk 7]
```diff
@@ -3943,10 +3943,14 @@ STAGE PLANS:
                 mode: mergepartial
                 outputColumnNames: _col0
                 Statistics: Num rows: 1 Data size: 4 Basic stats: COMPLETE Column stats: COMPLETE
-                Reduce Output Operator
-                  sort order: 
-                  Statistics: Num rows: 1 Data size: 4 Basic stats: COMPLETE Column stats: COMPLETE
-                  value expressions: _col0 (type: int)
+                Select Operator
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/llap/subquery_select.q.out]: Did the join change type due to this?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java [hunk 8]
```diff
@@ -1274,6 +1274,40 @@ protected ExprNodeDesc getXpathOrFuncExprNodeDesc(ASTNode expr,
               children);
         }
 
+        // introduce cast for expressions which are of different type than parent WHEN expr's type
```
Comment (t3rmin4t0r) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/TypeCheckProcFactory.java]: WHEN or THEN?

File: ql/src/test/results/clientpositive/perf/spark/query77.q.out [hunk 9]
```diff
@@ -826,7 +826,7 @@ STAGE PLANS:
                 outputColumnNames: _col1, _col2, _col3, _col4, _col5
                 Statistics: Num rows: 95833780 Data size: 13030622681 Basic stats: COMPLETE Column stats: NONE
                 Select Operator
-                  expressions: 'web channel' (type: string), _col3 (type: int), _col4 (type: decimal(17,2)), CASE WHEN (_col1 is not null) THEN (_col1) ELSE (0) END (type: decimal(17,2)), (_col5 - CASE WHEN (_col2 is not null) THEN (_col2) ELSE (0) END) (type: decimal(18,2))
+                  expressions: 'web channel' (type: string), _col3 (type: int), _col4 (type: decimal(17,2)), CASE WHEN (_col1 is not null) THEN (_col1) ELSE (CAST( 0 AS decimal(17,2))) END (type: decimal(17,2)), (_col5 - CASE WHEN (_col2 is not null) THEN (_col2) ELSE (CAST( 0 AS decimal(17,2))) END) (type: decimal(18,2))
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/perf/spark/query77.q.out]: No constant folding again?

File: ql/src/test/results/clientpositive/perf/spark/query80.q.out [hunk 10]
```diff
@@ -782,7 +782,7 @@ STAGE PLANS:
                     1 Map 33
                   Statistics: Num rows: 231917759 Data size: 31534108438 Basic stats: COMPLETE Column stats: NONE
                   Select Operator
-                    expressions: _col15 (type: string), _col5 (type: decimal(7,2)), CASE WHEN (_col9 is not null) THEN (_col9) ELSE (0) END (type: decimal(12,2)), (_col6 - CASE WHEN (_col10 is not null) THEN (_col10) ELSE (0) END) (type: decimal(13,2))
+                    expressions: _col15 (type: string), _col5 (type: decimal(7,2)), CASE WHEN (_col9 is not null) THEN (_col9) ELSE (CAST( 0 AS decimal(7,2))) END (type: decimal(7,2)), (_col6 - CASE WHEN (_col10 is not null) THEN (_col10) ELSE (CAST( 0 AS decimal(7,2))) END) (type: decimal(8,2))
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/perf/spark/query80.q.out]: Interesting case change, why is it not 12,2 or 13,2?

Comment (t3rmin4t0r) [reply ql/src/test/results/clientpositive/perf/spark/query80.q.out]: Also, inside it is 7,2, outside it is 8,2?

File: ql/src/test/results/clientpositive/spark/vectorized_case.q.out [hunk 11]
```diff
@@ -688,13 +688,13 @@ STAGE PLANS:
                       native: true
                       vectorizationSchemaColumns: [0:member:decimal(10,0)/DECIMAL_64, 1:attr:decimal(10,0)/DECIMAL_64, 2:ROW__ID:struct<writeid:bigint,bucketid:int,rowid:bigint>]
                   Select Operator
-                    expressions: CASE WHEN ((member = 1)) THEN (1) ELSE ((attr + 2)) END (type: decimal(11,0))
+                    expressions: CASE WHEN ((member = 1)) THEN (CAST( 1 AS decimal(11,0))) ELSE ((attr + 2)) END (type: decimal(11,0))
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/spark/vectorized_case.q.out]: Missed constant fold again?

File: ql/src/test/results/clientpositive/vectorized_case.q.out [hunk 12]
```diff
@@ -705,13 +705,13 @@ STAGE PLANS:
...
               Select Vectorization:
                   className: VectorSelectOperator
                   native: true
-                  projectedOutputColumnNums: [9]
-                  selectExpressions: IfExprDecimalColumnColumn(col 6:boolean, col 10:decimal(11,0)col 8:decimal(11,0))(children: Decimal64ColEqualDecimal64Scalar(col 0:decimal(10,0)/DECIMAL_64, decimal64Val 1, decimalVal 1) -> 6:boolean, ConvertDecimal64ToDecimal(col 7:decimal(11,0)/DECIMAL_64)(children: Decimal64ColAddDecimal64Scalar(col 1:decimal(10,0)/DECIMAL_64, decimal64Val 1, decimalVal 1) -> 7:decimal(11,0)/DECIMAL_64) -> 10:decimal(11,0), ConstantVectorExpression(val 2) -> 8:decimal(11,0)) -> 9:decimal(11,0)
+                  projectedOutputColumnNums: [8]
+                  selectExpressions: IfExprDecimal64ColumnDecimal64Scalar(col 6:boolean, col 7:decimal(11,0)/DECIMAL_64, decimal64Val 2, decimalVal 2)(children: Decimal64ColEqualDecimal64Scalar(col 0:decimal(10,0)/DECIMAL_64, decimal64Val 1, decimalVal 1) -> 6:boolean, Decimal64ColAddDecimal64Scalar(col 1:decimal(10,0)/DECIMAL_64, decimal64Val 1, decimalVal 1) -> 7:decimal(11,0)/DECIMAL_64) -> 8:decimal(11,0)/DECIMAL_64
```
Comment (t3rmin4t0r) [inline ql/src/test/results/clientpositive/vectorized_case.q.out]: This is pretty neat, the expression optimized to a faster one.

File: ql/src/test/results/clientpositive/perf/spark/query77.q.out (hunk 9 above)
Comment (vineetgarg02) [reply ql/src/test/results/clientpositive/perf/spark/query77.q.out]: This is due to calcite planner introducing case expression with mismatched type. Hive planner introduces cast to fix this and there is no (FULL) constant propagation afterwards. I'll do a follow-up to investigate and properly fix this.

File: ql/src/test/results/clientpositive/llap/subquery_select.q.out (hunk 7 above)
Comment (vineetgarg02) [reply ql/src/test/results/clientpositive/llap/subquery_select.q.out]: I don't see the join type being changed. The change only introduced an additional SELECT op. Can you point me to it?

File: ql/src/test/results/clientpositive/perf/spark/query80.q.out (hunk 10 above)
Comment (vineetgarg02) [reply ql/src/test/results/clientpositive/perf/spark/query80.q.out]: change to 7,2 is due to the `_col9`'s type which is decimal(7,2). The whole expression's type is determined to be decimal(8,2) for some reason, but this change is in line with previous logic where even though internal type was decimal(12,2) whole case expr was decimal(13,2). In short type inference looks to be more tight now.

File: ql/src/test/results/clientpositive/llap/multi_count_distinct_null.q.out (hunk 6 above)
Comment (vineetgarg02) [reply ql/src/test/results/clientpositive/llap/multi_count_distinct_null.q.out]: Latest update fixes constant folding issue.

**Pull request closed without merging**